### PR TITLE
feat(NXP-000): Adding special variant to cozyMarkers

### PR DESCRIPTION
**Pull request by pietroid**
## Description

Adding special variant to `cozyMarkers`:

image

This variant is needed for the favorite icon coloring:

image

### CODE REVIEW

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java [hunk 1]
```diff
@@ -57,11 +58,12 @@ private Path addPointerOnMarkerCenter(Bitmap marker, int pointerWidth, int point
         return pointer;
     }
 
-    private Bitmap getIconBitmap(String svgIcon, int width, int height) {
-        if(svgIcon == null)
+    private Bitmap getIconBitmap(String svgIcon, int width, int height, int rgbColor) {
+        String key = svgIcon + width + height + rgbColor;
+        if(key == null)
```
Comment (PankovSerge) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java]:
```suggestion
        if (key == null)
```

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java [hunk 2]
```diff
@@ -57,11 +58,12 @@ private Path addPointerOnMarkerCenter(Bitmap marker, int pointerWidth, int point
         return pointer;
     }
 
-    private Bitmap getIconBitmap(String svgIcon, int width, int height) {
-        if(svgIcon == null)
+    private Bitmap getIconBitmap(String svgIcon, int width, int height, int rgbColor) {
+        String key = svgIcon + width + height + rgbColor;
```
Comment (PankovSerge) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java]: Can we switch to hash string here?

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java [hunk 3]
```diff
@@ -77,6 +79,10 @@ private Bitmap getIconBitmap(String svgIcon, int width, int height) {
             String finalSvgIcon = svgIcon.replaceAll(widthPattern, "");
             finalSvgIcon = finalSvgIcon.replaceAll(heightPattern, "");
 
+            //Recoloring svg
+            String hexColor = String.format("#%06X", (0xFFFFFF & rgbColor));
+            finalSvgIcon = finalSvgIcon.replaceAll("(fill=\")(.+?)(\")", "fill=\""+hexColor+"\"");
+
```
Comment (PankovSerge) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java]: WDYT about extracting this part to a separate method?

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java [hunk 4]
```diff
@@ -124,6 +135,16 @@ private Bitmap getMarkerBitmap(CozyMarkerData markerData) {
...
+        if(markerData.variant.equals("special")){
+            if(markerData.isVisualized){
+                iconCircleColor = visualizedIconCircleColor;
+                iconColor = defaultIconColor;
+            }
+            if(markerData.isSelected) {
+                iconCircleColor = specialIconCircleColor;
+                iconColor = specialIconColor;
+            }
+        }
```
Comment (PankovSerge) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java]:
```suggestion
        if (markerData.variant.equals("special")){
            if (markerData.isVisualized){
                iconCircleColor = visualizedIconCircleColor;
                iconColor = defaultIconColor;
            }
            if (markerData.isSelected) {
                iconCircleColor = specialIconCircleColor;
                iconColor = specialIconColor;
            }
        }
```

File: packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/CozyMarkerBuilder.m [hunk 5]
```diff
@@ -61,16 +61,27 @@ void CFSafeRelease(CFTypeRef cf) {
...
     if(cachedImage != nil) {
         return cachedImage;
     }
     
     SVGKImage *svgImage = [SVGKImage imageWithSource:[SVGKSourceString sourceFromContentsOfString:svgIcon]];
     svgImage.size = CGSizeMake(width, height);
     UIImage* svgImageUI = svgImage.UIImage;
-    [[self cache] setObject:svgImageUI forKey:svgIcon];
+
+    if(color != nil){
```
Comment (PankovSerge) [inline packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/CozyMarkerBuilder.m]:
```suggestion
    if (color != nil){
```

File: packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/CozyMarkerBuilder.m [hunk 6]
```diff
@@ -197,14 +223,14 @@ - (UIImage *)getMarkerWithData:(CozyMarkerData *)cozyMarkerData {
         [text drawInRect:CGRectIntegral(textRect) withAttributes:@{NSFontAttributeName:textFont, NSForegroundColorAttributeName: textColor}];
         
         if(icon != NULL){
```
Comment (PankovSerge) [inline packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/CozyMarkerBuilder.m]:
```suggestion
        if (icon != NULL){
```

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java [hunk 7]
```diff
@@ -57,11 +58,16 @@ private Path addPointerOnMarkerCenter(Bitmap marker, int pointerWidth, int point
...
     }
 
-    private Bitmap getIconBitmap(String svgIcon, int width, int height) {
-        if(svgIcon == null)
+    private Bitmap getIconBitmap(String svgIcon, int width, int height, int rgbColor) {
+        if (svgIcon == null)
             return null;
         
-        final Bitmap bitmap = markerCache.getBitmapFromMemCache(svgIcon);   
+        String key = "" + svgIcon.hashCode() + width + height + rgbColor;
```
Comment (fulpismo) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java]: not a necessary suggestion, but would something like this work?
```suggestion
        String key = String.format("%d%d%d%d", svgIcon.hashCode(), width, height, rgbColor);
```

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java [hunk 8]
```diff
@@ -88,11 +96,16 @@ private Bitmap getIconBitmap(String svgIcon, int width, int height) {
 
             markerCache.addBitmapToMemoryCache(svgIcon, iconBitmap);
             return iconBitmap;
-        }catch(Exception e){}
+        }catch (Exception e) {}
 
         return null;
     }
 
+    private String recolorSvg (String originalSvg, int rgbColor) {
```
Comment (fulpismo) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java]:
```suggestion
    private String recolorSvg(String originalSvg, int rgbColor) {
```

**Comment by fulpismo** [comment]
the print in description is for android or iOS? is there any difference in the design?

**Comment by pietroid** [comment]
> the print in description is for android or iOS? is there any difference in the design?

There should be no difference! The print is on Android

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java (hunk 7 above)
Comment (pietroid) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java]: Thanks!

**Review by fulpismo — APPROVED** [review]

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java (hunk 7 above)
Comment (pietroid) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/main/java/io/flutter/plugins/googlemaps/CozyMarkerBuilder.java]: Applied!

**Pull request merged**